internal/controller/healthchecker.go

```diff
@@ -0,0 +1,161 @@
+)
+
+type HealthChecker struct {
+	client.Reader    // intentional, we do not want to use the cache
+	Update           chan event.GenericEvent
+	Logger           log.Logger
+	NodeName         string
+	ApplyEmptyConfig EmptyConfigApplier
+	ReloadStatus     func()
+	ctx              context.Context
+	healthResult     string
+	healthResMutex   sync.Mutex
+	check            healthcheck
+	attempts         int
+	interval         time.Duration
+}
+
+type healthcheck func() error
+
+func (r *HealthChecker) HealthResult() string {
+	r.healthResMutex.Lock()
+	defer r.healthResMutex.Unlock()
+	return r.healthResult
+}
+
+type EmptyConfigApplier interface {
+	ApplyEmptyConfig() error
+}
+
+type hcEvent struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+}
+
+func (evt *hcEvent) DeepCopyObject() runtime.Object {
+	res := new(hcEvent)
+	res.Name = evt.Name
+	res.Namespace = evt.Namespace
+	return res
+}
+
+func NewHealthCheckEvent() event.GenericEvent {
+	evt := hcEvent{}
+	evt.Namespace = "healthcheck"
+	evt.Name = "check"
+	return event.GenericEvent{Object: &evt}
+}
+
+func NewHealthChecker(cl client.Reader, update chan event.GenericEvent, l log.Logger,
+	node string, apply EmptyConfigApplier, reloadStatus func(), ctx context.Context,
+	attempts int, timeout, interval time.Duration) *HealthChecker {
+	return &HealthChecker{
+		Reader:           cl,
+		Update:           update,
+		Logger:           l,
+		NodeName:         node,
+		ApplyEmptyConfig: apply,
+		ReloadStatus:     reloadStatus,
+		ctx:              ctx,
+		check:            healthcheckForAPIServer(cl, ctx, timeout, node),
+		attempts:         attempts,
+		interval:         interval,
+	}
+}
+
+func (r *HealthChecker) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "HealthChecker", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "HealthChecker", "end reconcile", req.NamespacedName.String())
+
+	lastHealthResult := r.healthResult
+	healthResult := r.healthResult
+
+	defer func() {
+		r.healthResMutex.Lock()
+		r.healthResult = healthResult
+		r.healthResMutex.Unlock()
+		if healthResult != lastHealthResult {
+			r.ReloadStatus()
+		}
+	}()
+
+	var err error
+	for i := 0; i < r.attempts; i++ {
+		err = r.check()
+		if err == nil {
+			break
+		}
+	}
+	if err == nil {
+		return ctrl.Result{}, nil
+	}
+	level.Error(r.Logger).Log("controller", "HealthChecker", "failed, error: ", err)
+
+	err = r.ApplyEmptyConfig.ApplyEmptyConfig()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	healthResult = time.Now().Format(time.DateTime)
+
+	configStale.Set(1) // the config controller would set it 0 the next time a valid config is applied
+	configReset.Inc()
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HealthChecker) SetupWithManager(mgr ctrl.Manager) error {
+	if r.interval > 0 {
+		go r.feedUpdate()
```

Comment:
we should pass the outer context here so the goroutine can stop

Reply:
it uses the context that is passed to the manager (`r.ctx.Done()`) to stop, do you think I should change it so it is  part of the Setup instead of a field on the reconciler?